### ISSUE TITLE
[BUG] possible memory corruption when multiplying large matices

### ISSUE THREAD

**Issue by vojirt** [tensorflow/tensorflow#22603]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**:
binary (through pip)
- **TensorFlow version (use command below)**:
1.11.0
- **Python version**:
2.7.12
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
9.1.85
- **GPU model and memory**:
TITAN Xp 12GB
- **Exact command to reproduce**:

more system information in [tf_env.txt](https://github.com/tensorflow/tensorflow/files/2429099/tf_env.txt)

### Describe the problem
For large matrix multiplications, part of the result is not correctly computed or its memory is partially overwritten. Not sure what exactly the problem is. I created as small as possible reproducible code that can be run and visualise the problem. 

Here are also visual outputs that I run. The result should be array of constant value 3.
with tf.tile:
sampling = 6
![sampling = 6](https://user-images.githubusercontent.com/8818326/46223432-45741880-c34b-11e8-8d0e-88c4a4682073.png)
sampling = 8
![sampling = 8](https://user-images.githubusercontent.com/8818326/46223453-591f7f00-c34b-11e8-9fcc-6edf42434fc3.png)
with broadcasting:
sampling = 6
![sampling = 6](https://user-images.githubusercontent.com/8818326/46223487-73595d00-c34b-11e8-9314-179848bb2f96.png)
sampling = 8
![sampling = 8](https://user-images.githubusercontent.com/8818326/46223500-7b190180-c34b-11e8-9122-31fb2e0991c7.png)

### Source code / logs
by varying the sampling variable from 6 to 12 (OOM) you can observe the behaviour by plotting the result (as done in the code)

```python
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
from mpl_toolkits.axes_grid1 import make_axes_locatable


# On Nvidia Titan XP 12GB:
# 1) MEMORY BUG - for shape [960, 480] and sampling > 6 the output 
#                 variable is partially overwritten (see figure) 
# 2) OOM error when sampling >= 12
sampling = 8  
shape = [480, 960]

# "random" input vectors
i1 = np.ones(shape=(shape[0]*shape[1], 1, 3), dtype=np.float)
i2 = np.ones(shape=(1, sampling**3, 3), dtype=np.float)

input1 = tf.placeholder(shape=(i1.shape[0], i1.shape[1], i1.shape[2]), dtype=tf.float32)
input2 = tf.placeholder(shape=(i2.shape[0], i2.shape[1], i2.shape[2]), dtype=tf.float32)

# Operation v1:  
# tile before multiplying to have same dimensions
# vectors (batch size, shape[0]*shape[1], sampling^2 , 3) and
# to get (after summation) vector (barch size, shape[0]*shape[1], sampling^3)
i1_tile = tf.tile(input1, (1, sampling**3, 1))
i2_tile = tf.tile(input2, (shape[0]*shape[1], 1, 1))
res1 = tf.reduce_sum(i1_tile * i2_tile, axis=-1)

# Operation v2:  
# use broadcasting to multiply 
# vectors i1 = (batch size, shape[0]*shape[1],     1     , 3) and
#         i2 = (batch size,         1        , sampling^3, 3) 
# to get (after summation) vector (barch size, shape[0]*shape[1], sampling^3)
res2 = tf.reduce_sum(input1*input2, axis=-1)

res = res1 #res2

with tf.Session() as sess:
    pred_out = sess.run(res, feed_dict={input1: i1, input2: i2})

# result should be image of size shape with all values equal to 3
fig = plt.figure()
pred_out = np.reshape(np.mean(pred_out, axis=-1), (shape[0], shape[1]))
imgplot = plt.imshow(pred_out)
imgplot.set_cmap('jet')
divider = make_axes_locatable(plt.gca())
cax1 = divider.append_axes("right", size="5%", pad=0.05)
fig.colorbar(imgplot, cax=cax1)
plt.clim(0, 3.1)

plt.savefig("result.png", dpi=100)

```

**Console output of the run of the script:**
/home/mifs/tv278/.virtualenvs/tf-keras/local/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
2018-09-28 18:22:45.596062: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2018-09-28 18:22:45.865097: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1411] Found device 0 with properties: 
name: TITAN Xp major: 6 minor: 1 memoryClockRate(GHz): 1.582
pciBusID: 0000:17:00.0
totalMemory: 11.91GiB freeMemory: 11.74GiB
2018-09-28 18:22:45.865123: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1490] Adding visible gpu devices: 0
2018-09-28 18:22:46.099514: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-09-28 18:22:46.099541: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977]      0 
2018-09-28 18:22:46.099546: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990] 0:   N 
2018-09-28 18:22:46.099791: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1103] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 11359 MB memory) -> physical GPU (device: 0, name: TITAN Xp, pci bus id: 0000:17:00.0, compute capability: 6.1)

**Comment by unrealwill** [comment]
probably related to  #19155

**Comment by vojirt** [comment]
Seems like the same problem as #19155

**Comment by tensorflowbutler** [comment]
Nagging Assignees @chsigg, @azaks2: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @chsigg, @azaks2: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by chsigg** [comment]
I've reproduced the issue. The problem seems to be with the reduction, not with the matrix multiplication. I'm closing this issue and will provide updates in #22123.
